samba4: update to 4.9.5, switch to embedded heimdal build

Andy2244 opened the pull request:
Maintainer: me
Compile tested: arm/mips (snapshots)
Run tested: arm, mips(qemu) (snapshots)

Description:
* update to 4.9.5
* use embedded heimdal build
* remove mitkrb5 deps
* add AD-DC tools (samba4kinit samba4kgetcred samba4kpasswd samba4ktutil)
* add eventlogadm to ADMIN tools
* build VFS modules shared
* preserve custom smb.conf after reboots
* add missing lib/samba/? private dirs, even if empty
* handle python-crypto via EXTRA_DEPENDS
* fix CONFIGURE_CMD argument order
* move python install to libs
* remove redundant Config.in selects

NOTE: MIT Kerberos support was relegated to experimental status and it seems the samba.org team does not want to support it in the future, so i switched to the embedded Heimdal build option. 
I picked this option over adding a new separate external Heimdal package, to avoid a similar problem in the future.

@val-kulkov I remember you asking for this option, so maybe you can also test this version with your setup.

val-kulkov commented:
@Andy2244 : I am very impressed that you actually managed to tackle this beast. Thank you. This is a major undertaking, and one that is important to me. I will get to testing this PR on all three of my platforms as soon as I return home, which will be toward the end of next week.

val-kulkov commented:
Found one problem so far when installing samba4-client on a system where libtalloc is already present:

```
... <snip> ...
Configuring libgnutls.
Configuring libtirpc.
Collected errors:
 * check_data_file_clashes: Package samba4-libs wants to install file /usr/lib/libtalloc.so.2
	But that file is already provided by package  * libtalloc
 * opkg_install_cmd: Cannot install package samba4-client.
```

libtalloc is installed by freeradius3. If libtalloc and freeradius3 are removed (for testing samba4 only), samba4-client installs successfully.

Now run-testing samba4 client and server; will report test results a little later.

val-kulkov commented:
@Andy2244 : provisioning an AD with samba-tool fails, apparently due to missing Python modules:
```
root@bas16dev:~# samba-tool domain provision --use-rfc2307 --interactive --option="password hash userPassword schemes = CryptSHA256"
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Traceback (most recent call last):
  File "/usr/bin/samba-tool", line 33, in <module>
    from samba.netcmd.main import cmd_sambatool
  File "/usr/lib/python2.7/site-packages/samba/__init__.py", line 27, in <module>
    import time
ImportError: No module named time
```
I very briefly checked OpenWrt python modules and concluded that it would take me considerable time to identify what provides the missing module. Leaving that to another day, I switched on to testing domain join as a domain member to an existing AD. The domain join succeeded:
```
root@bas16dev:~# net ads join -U administrator
Enter administrator's password:
ldb: unable to stat module /usr/lib/samba/ldb : No such file or directory
Using short domain name -- HOME
Joined 'BAS16DEV' to dns domain 'home.lan'
DNS update failed: NT_STATUS_UNSUCCESSFUL
```
The DNS update failure is likely a misconfiguration on the client; I have seen it before and I would not worry about that. The "ldb: unable to stat module" is a problem on the server. In /usr/lib/samba, I only have two subdirectories, idmap and krb5. There should be also auth, vfs and pdb, but they are missing because these subdirectories are only checked for existence, but not created, in Makefile's "Package/samba4-server/install" section. They must be created.

I would also suggest that the test for symlink on line 62 in /etc/init.d/samba4 is replaced with a test for existence of a file, such as:
```
   [ -e /etc/samba/smb.conf ] || ln -nsf /var/etc/smb.conf /etc/samba/smb.conf
```
While it is convenient to have an initial version of smb.conf created automatically from the template and settings in /etc/config/samba4, it is impossible at present to specify directives in /etc/config/samba4 that are not recognized by the init script. To specify those other directives, I replaced the symlink with an actual file, /etc/samba/smb.conf. With the current test in the init script, "-L /etc/samba/smb.conf", such file gets removed on each Samba restart and is replaced with a symlink. Not good.

Andy2244 commented:
ok thanks will try to reproduce this and fix those problems.

Andy2244 commented:
@val-kulkov ok still working on this one, but can you check if python-base was installed correctly?
Whats the output if you use the interactive command-line?

```
python
import time
```

There should be a `time.so` in `/usr/lib/python2.7/lib-dynload`, if its not there can you try install the python base package and python-crypto?

val-kulkov commented:
Hi @Andy2244 , it looks like time.so comes with "python" and/or its dependencies, and not "python-base":

```
root@bas16dev:~# opkg list-installed | grep python
python-base - 2.7.16-1
root@bas16dev:~# find /usr/lib/python2.7/ -name "time*"
/usr/lib/python2.7/site-packages/samba/tests/samba_tool/timecmd.py
root@bas16dev:~#
```
Then, "opkg -autoremove remove samba4-admin samba4-server samba4-client python-base", and

```
root@bas16dev:~# opkg install python
Installing python (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python_2.7.16-1_arm_cortex-a9.ipk
Installing python-base (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-base_2.7.16-1_arm_cortex-a9.ipk
Installing libffi (3.2.1-3) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/libffi_3.2.1-3_arm_cortex-a9.ipk
Installing python-light (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-light_2.7.16-1_arm_cortex-a9.ipk
Installing python-ctypes (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-ctypes_2.7.16-1_arm_cortex-a9.ipk
Installing python-pydoc (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-pydoc_2.7.16-1_arm_cortex-a9.ipk
Installing python-multiprocessing (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-multiprocessing_2.7.16-1_arm_cortex-a9.ipk
Installing python-logging (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-logging_2.7.16-1_arm_cortex-a9.ipk
Installing python-codecs (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-codecs_2.7.16-1_arm_cortex-a9.ipk
Installing python-sqlite3 (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-sqlite3_2.7.16-1_arm_cortex-a9.ipk
Installing python-xml (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-xml_2.7.16-1_arm_cortex-a9.ipk
Installing libgdbm (1.11-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/libgdbm_1.11-1_arm_cortex-a9.ipk
Installing python-gdbm (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-gdbm_2.7.16-1_arm_cortex-a9.ipk
Installing python-email (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-email_2.7.16-1_arm_cortex-a9.ipk
Installing python-distutils (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-distutils_2.7.16-1_arm_cortex-a9.ipk
Installing python-openssl (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-openssl_2.7.16-1_arm_cortex-a9.ipk
Installing python-compiler (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-compiler_2.7.16-1_arm_cortex-a9.ipk
Installing python-db (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-db_2.7.16-1_arm_cortex-a9.ipk
Installing python-decimal (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-decimal_2.7.16-1_arm_cortex-a9.ipk
Installing python-unittest (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-unittest_2.7.16-1_arm_cortex-a9.ipk
Installing python-ncurses (2.7.16-1) to root...
Downloading http://vk2011:6554/bcm53xx/bin/packages/arm_cortex-a9/packages/python-ncurses_2.7.16-1_arm_cortex-a9.ipk
Configuring python-base.
Configuring libffi.
Configuring python-light.
Configuring python-codecs.
Configuring python-db.
Configuring python-decimal.
Configuring python-distutils.
Configuring python-pydoc.
Configuring python-ctypes.
Configuring python-multiprocessing.
Configuring python-logging.
Configuring python-sqlite3.
Configuring python-xml.
Configuring libgdbm.
Configuring python-gdbm.
Configuring python-email.
Configuring python-openssl.
Configuring python-compiler.
Configuring python-unittest.
Configuring python-ncurses.
Configuring python.
root@bas16dev:~# find /usr/lib/python2.7/ -name "time*"
/usr/lib/python2.7/lib-dynload/time.so
/usr/lib/python2.7/timeit.pyc
root@bas16dev:~# python
Python 2.7.16 (default, Mar 18 2019, 19:55:52) 
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> 
root@bas16dev:~# 
```

Well, that does not exactly answer the question where time.so comes from in this particular case. So, here:
```
val@vk2011:~/Development/openwrt/targets/bcm53xx$ for f in bin/packages/arm_cortex-a9/packages/python*ipk ; do tar xzf $f ./data.tar.gz ; tar tvf data.tar.gz | grep time.so && echo $f ; rm data.tar.gz ; done
-rwxr-xr-x root/root      4134 2019-03-18 15:55 ./usr/lib/python2.7/site-packages/cryptography/hazmat/bindings/_constant_time.so
bin/packages/arm_cortex-a9/packages/python-cryptography_2.6.1-1_arm_cortex-a9.ipk
-rwxr-xr-x root/root     60910 2019-03-18 15:55 ./usr/lib/python2.7/lib-dynload/datetime.so
-rwxr-xr-x root/root     16314 2019-03-18 15:55 ./usr/lib/python2.7/lib-dynload/time.so
bin/packages/arm_cortex-a9/packages/python-light_2.7.16-1_arm_cortex-a9.ipk
```

So the answer is that time.so comes from python-light. Not from python-base. I hope this helps.

Andy2244 commented:
@val-kulkov ok found the problem, actually `python-crypto` will select a whole bunch of python packages, including python-light. Yet because i had this buildroot configuration bug, i could only add the dependency via `PKG_BUILD_DEPENDS` and as select logic via menuconfig.

I suspect you manually added the updated package and did not build a full new fw via menuconfig, so you seem to miss the `python-crypto` package and all its selected python dependencies.

I will try to figure out if there is a way to resolve the `recursive dependency detected!` error if i try add `+SAMBA4_SERVER_AD_DC:python-crypto`.

Meanwhile just install python-crypto, which should install all needed modules.

Andy2244 commented:
ok found the solution, thanks to the obfsproxy makefile. It seems you need to always add the `+python-light` along with `+python-crypto` or you run into this `recursive dependency detected!` error.

Andy2244 commented:
@val-kulkov ok i reworked the python handling and also fixed the other issues. Your local smb.conf is now also preserved.

Let me know if this fixes your issues.

Andy2244 commented:
No clue why the download fails for CircleCI...? Works for me locally.

neheb commented:
curl: (16) Error in the HTTP2 framing layer

weird error.

Andy2244 commented:
@diizzyy @neheb Anyone up for reviewing this? (Just ignore the ad-dc stuff for now)

diizzyy commented on net/samba4/Makefile:
```diff
@@ -303,15 +369,16 @@ define Build/Configure
-		./buildtools/bin/waf install -j$(shell nproc) \
-		--targets=$(SAMBA4_MODULES)$(SAMBA4_MODULES_SHARDED)$(BUILD_TARGETS_SERVER),$(BUILD_TARGETS_UTILS),$(BUILD_TARGETS_ADMIN),$(BUILD_TARGETS_CLIENT) \
+		./buildtools/bin/waf install \
+		--jobs=$(shell nproc) \
```
This should use to the value set by toolchain, not blindly override it?

Andy2244 replied:
If you can tell me how to get this value? I could not find any variable that stores this.

diizzyy replied:
https://github.com/openwrt/openwrt/blob/master/tools/cmake/Makefile#L23
A bit short on time right now but it looks like cmake uses this?

Andy2244 replied:
ok thanks, found it:
PKG_JOBS?=$(if $(PKG_BUILD_PARALLEL),$(MAKE_J),-j1)

Andy2244 replied:
Yeah i remember, i tried this way back and this is not working. 
What i get is: `waf install --jobserver-auth=3,4 ` instead of `waf install -j4` via `waf install $(PKG_JOBS)`
`make -j8` gives me `waf: error: no such option: --jobserver-auth`

So i don't see a easy way to get the make jobs?

diizzyy replied:
Yeah, doesn't like there's a clean way of doing it :-(

diizzyy commented on net/samba4/Makefile:
```diff
@@ -2,19 +2,18 @@
 PKG_LICENSE:=GPL-3.0-only
 PKG_LICENSE_FILES:=COPYING
 
 PKG_SOURCE_URL:=https://download.samba.org/pub/samba/stable/
```
As a workaround for perhaps we can add the following mirrors to at least try to get a clean CI build?
https://ftp5.gwdg.de/pub/samba/stable/
https://ftp.yz.yamagata-u.ac.jp/pub/network/samba/
http://ftp.uni-bayreuth.de/netsoftware/samba/

Andy2244 replied:
sure np

diizzyy commented:
Fails to build :-/

```
        Selected embedded Heimdal build
Waf: Entering directory `/home/diizzy/linksys/openwrt/build_dir/hostpkg/samba-4.9.5/bin'
        Selected embedded Heimdal build
[161/221] Generating nfs41acl-h
/bin/sh: rpcgen: not found
Waf: Leaving directory `/home/diizzy/linksys/openwrt/build_dir/hostpkg/samba-4.9.5/bin'
Build failed:  -> task failed (err #127):
        {task: nfs41acl-h nfs41acl.x -> nfs41acl.h}
make[3]: *** [Makefile:471: /home/diizzy/linksys/openwrt/build_dir/hostpkg/samba-4.9.5/.built] Error 1
```

Andy2244 commented:
@diizzyy ok fixed, forgot why i had `nfs-kernel-server/host` there...

diizzyy commented:
Merging this at the end of the day unless there are any objects
